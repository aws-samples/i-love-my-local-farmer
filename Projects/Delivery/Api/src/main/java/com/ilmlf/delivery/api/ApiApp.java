package com.ilmlf.delivery.api;

import java.io.IOException;
import software.amazon.awscdk.core.App;
import software.amazon.awscdk.core.Environment;

public class ApiApp {
  public static void main(final String[] args) throws IOException {
    App app = new App();

    new ApiStack(
        app,
        "ApiStack",
        ApiStack.ApiStackProps.builder()
            .env(Environment.builder()
                .account("433621526002")
                .region("eu-west-1")
                .build())
            .dbEndpoint("farmerdb.cyo3qc04zrsp.eu-west-1.rds.amazonaws.com:3306")
            .dbProxyEndpoint(
                "farmerdbproxy.proxy-cyo3qc04zrsp.eu-west-1.rds.amazonaws.com:3306/FarmerDB")
            .dbProxyArn("arn:aws:rds:eu-west-1:433621526002:db-proxy:prx-0a792ad6c6e7aef99")
            .dbRegion("eu-west-1")
            .dbUser("lambda_iam")
            .dbUserSecretName("DbUserSecret")
            .dbUserSecretArn("arn:aws:secretsmanager:eu-west-1:433621526002:secret:DbUserSecret-CeVI0L")
            .dbAdminSecretName("DeliveryProjectDbFarmerDBSe-2OOKfJWB3Ue2")
            .dbAdminSecretArn(
                "arn:aws:secretsmanager:eu-west-1:433621526002:secret:DeliveryProjectDbFarmerDBSe-2OOKfJWB3Ue2-mRuwvG")
            .dbSgId("sg-09e6e9b4696bb2a53")
            .dbVpcId("vpc-0505198ab2080d218")
            .build());

    app.synth();
  }
}
