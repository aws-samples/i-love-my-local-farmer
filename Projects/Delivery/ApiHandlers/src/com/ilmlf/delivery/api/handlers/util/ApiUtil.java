package com.ilmlf.delivery.api.handlers.util;

import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyResponseEvent;

public abstract class ApiUtil {

  public static APIGatewayProxyResponseEvent generateReturnData(Integer httpStatus, String message) {
    APIGatewayProxyResponseEvent response = new APIGatewayProxyResponseEvent();
    response.setStatusCode(httpStatus);
    response.setBody(message);
    return response;
  }
}
