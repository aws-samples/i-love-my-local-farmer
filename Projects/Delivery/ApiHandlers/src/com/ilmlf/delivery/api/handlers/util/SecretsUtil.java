/*
Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
Licensed under the Apache License, Version 2.0 (the "License").
You may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package com.ilmlf.delivery.api.handlers.util;

import org.json.JSONObject;
import lombok.NonNull;
import software.amazon.awssdk.http.urlconnection.UrlConnectionHttpClient;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.secretsmanager.SecretsManagerClient;
import software.amazon.awssdk.services.secretsmanager.model.GetSecretValueRequest;
import software.amazon.awssdk.services.secretsmanager.model.GetSecretValueResponse;
import software.amazon.awssdk.services.secretsmanager.model.SecretsManagerException;

public abstract class SecretsUtil {

  public static JSONObject getSecret(@NonNull String region, @NonNull String dbSecretName) {
    SecretsManagerClient secretsClient = SecretsManagerClient.builder()
            .region(Region.of(region))
            .httpClientBuilder(UrlConnectionHttpClient.builder())
            .build();

    try {
      GetSecretValueRequest valueRequest = GetSecretValueRequest.builder()
              .secretId(dbSecretName)
              .build();

      GetSecretValueResponse valueResponse = secretsClient.getSecretValue(valueRequest);
      String secret = valueResponse.secretString();
      JSONObject jo = new JSONObject(secret);
      secretsClient.close();
      return jo;
    } catch (SecretsManagerException e) {
      System.err.println(e.awsErrorDetails().errorMessage());
      throw new RuntimeException(e);
    }
  }
}
