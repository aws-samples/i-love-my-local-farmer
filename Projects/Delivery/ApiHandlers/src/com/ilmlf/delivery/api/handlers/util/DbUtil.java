/*
Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
Licensed under the Apache License, Version 2.0 (the "License").
You may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package com.ilmlf.delivery.api.handlers.util;

import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.net.URL;
import java.security.KeyStore;
import java.security.cert.CertificateFactory;
import java.security.cert.X509Certificate;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.util.Properties;
import lombok.NonNull;
import software.amazon.awssdk.auth.credentials.DefaultCredentialsProvider;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.rds.RdsUtilities;
import software.amazon.awssdk.services.rds.model.GenerateAuthenticationTokenRequest;

public abstract class DbUtil {
  private static final String SSL_CERTIFICATE = "rds-ca-2019-root.pem";
  private static final String KEY_STORE_TYPE = "JKS";
  private static final String KEY_STORE_PROVIDER = "SUN";
  private static final String KEY_STORE_FILE_PREFIX = "sys-connect-via-ssl-test-cacerts";
  private static final String KEY_STORE_FILE_SUFFIX = ".jks";
  private static final String DEFAULT_KEY_STORE_PASSWORD = "delivery";
  private static final String JDBC_PREFIX = "jdbc:mysql://";

  /**
   * Create a database connection via a IAM Authentication. <br/>
   * The password will be generated via an Authentication Token using an rds cert.
   * @param username
   * @param dbEndpoint
   * @param region
   * @return
   */
  public static Connection createConnectionViaIamAuth(@NonNull String username, @NonNull String dbEndpoint, @NonNull String region) {
    Connection connection;
    try {
      setSslProperties();
      connection = DriverManager.getConnection(JDBC_PREFIX + dbEndpoint, setMySqlConnectionProperties(username, dbEndpoint, region));
      return connection;
      
    } catch (Exception e) {
      // TODO add retry-logic
      System.out.println("Connection FAILED");
      e.printStackTrace();
    }
    return null;
  }
  
  /**
   * Create a database connection via a username + password authentication
   * @param username
   * @param pwd
   * @param dbEndpoint
   * @return
   */
  public static Connection createConnectionViaUserPwd(@NonNull String username, @NonNull String pwd, @NonNull String dbEndpoint) {
    Connection connection;
    try {
      connection = DriverManager.getConnection(JDBC_PREFIX + dbEndpoint, username, pwd);
      System.out.println("Connection Established");
      return connection;
    } catch (SQLException e) {
      // TODO add retry-logic
      System.out.println("Connection FAILED");
      e.printStackTrace();
    }
    return null;
  }

  
  /**
   * This method generates the IAM Authentication Token, <br/>
   * which will be later used as the password for authenticating to the DB.
   * @return the authentication token
   */
  private static String generateAuthToken(String username, String dbEndpoint, String region){
    long start = System.currentTimeMillis();
    RdsUtilities utilities = RdsUtilities.builder()
      .credentialsProvider(DefaultCredentialsProvider.create())
      .region(Region.of(region))
      .build();
    GenerateAuthenticationTokenRequest authTokenRequest = GenerateAuthenticationTokenRequest.builder()
      .username(username)
      .hostname(dbEndpoint)
      .port(3306)
      .build();
    String authenticationToken = utilities.generateAuthenticationToken(authTokenRequest);

    return authenticationToken;
  }

  /**
   * This method sets the mysql connection properties which includes the IAM Database Authentication token
   * as the password. It also specifies that SSL verification is required.
   * @return
  */
  private static Properties setMySqlConnectionProperties(String username, String dbEndpoint, String region) {
    Properties mysqlConnectionProperties = new Properties();
    mysqlConnectionProperties.setProperty("useSSL", "true");
    mysqlConnectionProperties.setProperty("user", username);
    mysqlConnectionProperties.setProperty("password", generateAuthToken(username, dbEndpoint, region));
    return mysqlConnectionProperties;
  }
  
  /**
   * This method sets the SSL properties which specify the key store file, its type and password:
   * @throws Exception
  */
  private static void setSslProperties() throws Exception {
    File keyStoreFile = createKeyStoreFile(createCertificate());
    System.setProperty("javax.net.ssl.trustStore", keyStoreFile.getPath());
    System.setProperty("javax.net.ssl.trustStoreType", KEY_STORE_TYPE);
    System.setProperty("javax.net.ssl.trustStorePassword", DEFAULT_KEY_STORE_PASSWORD);
   }

  /**
   *  This method generates the SSL certificate
   * @return
   * @throws Exception
  */
  private static X509Certificate createCertificate() throws Exception {
    CertificateFactory certFactory = CertificateFactory.getInstance("X.509"); 
    URL url = new File(SSL_CERTIFICATE).toURI().toURL();
    try (InputStream certInputStream = url.openStream()) {
        return (X509Certificate) certFactory.generateCertificate(certInputStream);
    }
  }

  /**
   * This method creates the Key Store File needed for the SSL verification <br/>
   * during the IAM Database Authentication to the db instance
   * @param rootX509Certificate - the SSL certificate to be stored in the KeyStore
   * @return
   * @throws Exception
  */
  private static File createKeyStoreFile(X509Certificate rootX509Certificate) throws Exception {
    File keyStoreFile = File.createTempFile(KEY_STORE_FILE_PREFIX, KEY_STORE_FILE_SUFFIX);
    try (FileOutputStream fos = new FileOutputStream(keyStoreFile.getPath())) {
        KeyStore ks = KeyStore.getInstance(KEY_STORE_TYPE, KEY_STORE_PROVIDER);
        ks.load(null);
        ks.setCertificateEntry("rootCaCertificate", rootX509Certificate);
        ks.store(fos, DEFAULT_KEY_STORE_PASSWORD.toCharArray());
    }
    return keyStoreFile;
  }
}
