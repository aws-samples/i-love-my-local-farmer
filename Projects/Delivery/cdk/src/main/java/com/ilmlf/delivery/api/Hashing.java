/*
Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
Licensed under the Apache License, Version 2.0 (the "License").
You may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package com.ilmlf.delivery.api;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.SequenceInputStream;
import java.nio.file.Files;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.Vector;
import org.apache.commons.codec.digest.DigestUtils;

public class Hashing {
  public static String hashDirectory(String directoryPath, boolean includeHiddenFiles)
      throws IOException {
    File directory = new File(directoryPath);

    if (!directory.isDirectory()) {
      throw new IllegalArgumentException("Not a directory");
    }

    Vector<FileInputStream> fileStreams = new Vector<>();
    collectFiles(directory, fileStreams, includeHiddenFiles);

    try (SequenceInputStream sequenceInputStream =
        new SequenceInputStream(fileStreams.elements())) {
      return DigestUtils.md5Hex(sequenceInputStream);
    }
  }

  private static void collectFiles(
      File directory, List<FileInputStream> fileInputStreams, boolean includeHiddenFiles)
      throws IOException {
    File[] files = directory.listFiles();

    if (files != null) {
      Arrays.sort(files, Comparator.comparing(File::getName));

      for (File file : files) {
        if (includeHiddenFiles || !Files.isHidden(file.toPath())) {
          if (file.isDirectory()) {
            collectFiles(file, fileInputStreams, includeHiddenFiles);
          } else {
            fileInputStreams.add(new FileInputStream(file));
          }
        }
      }
    }
  }
}
