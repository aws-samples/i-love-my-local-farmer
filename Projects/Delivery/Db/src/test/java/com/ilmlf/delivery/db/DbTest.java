package com.ilmlf.delivery.db;

import io.github.jsonSnapshot.SnapshotMatcher;
import java.util.Arrays;
import java.util.Map;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import software.amazon.awscdk.core.App;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;

import java.io.IOException;

import org.junit.jupiter.api.Test;
import software.amazon.awscdk.core.AppProps;
import software.amazon.awscdk.core.Environment;
import software.amazon.awscdk.core.StackProps;

import static io.github.jsonSnapshot.SnapshotMatcher.expect;
import static org.assertj.core.api.Assertions.assertThat;

public class DbTest {
    private final static ObjectMapper JSON =
        new ObjectMapper()
            .configure(SerializationFeature.INDENT_OUTPUT, true)
            .configure(SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS, true);

    @BeforeAll
    public static void beforeAll() {
        SnapshotMatcher.start(Snapshot::asJsonString);
    }

    @AfterAll
    public static void afterAll() {
        SnapshotMatcher.validateSnapshots();
    }

    @Test
    public void testStack() {
        App app = new App();

        DbStack stack = new DbStack(app, "test");

        // synthesize the stack to a CloudFormation template
        JsonNode actual =
            JSON.valueToTree(app.synth().getStackArtifact(stack.getArtifactId()).getTemplate());

        expect(actual).toMatchSnapshot();

    }
}
