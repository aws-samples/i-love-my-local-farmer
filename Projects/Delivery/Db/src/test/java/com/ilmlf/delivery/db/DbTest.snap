com.ilmlf.delivery.db.DbTest.testStack=[
  {
    "Outputs": {
      "configJsonContents": {
        "Description": "Config file (in json format) for Stacks with Lambdas",
        "Value": {
          "Fn::Join": [
            "",
            [
              "{DB_PROXY_ARN=",
              {
                "Fn::GetAtt": [
                  "FarmerDbProxy7A9720C2",
                  "DBProxyArn"
                ]
              },
              ", DB_PROXY_ENDPOINT=",
              {
                "Fn::GetAtt": [
                  "FarmerDbProxy7A9720C2",
                  "Endpoint"
                ]
              },
              ":3306/FarmerDB, DB_REGION=",
              {
                "Ref": "AWS::Region"
              },
              ", DB_SECRET_ARN=",
              {
                "Ref": "FarmerDBSecretAttachment8D79E04A"
              },
              ", DB_SECRET_NAME=",
              {
                "Fn::Select": [
                  6,
                  {
                    "Fn::Split": [
                      ":",
                      {
                        "Ref": "testFarmerDBSecretCBD37E0A3fdaad7efa858a3daf9490cf0a702aeb"
                      }
                    ]
                  }
                ]
              },
              ", DB_SG_ID=",
              {
                "Fn::GetAtt": [
                  "FarmerDeliverySGDA50C27C",
                  "GroupId"
                ]
              },
              ", DB_VPC_ID=",
              {
                "Ref": "farmerVpc3E1EF418"
              },
              "}"
            ]
          ]
        }
      }
    },
    "Resources": {
      "FarmerDBC89642C8": {
        "DeletionPolicy": "Snapshot",
        "Properties": {
          "AllocatedStorage": "25",
          "AutoMinorVersionUpgrade": true,
          "BackupRetentionPeriod": 7,
          "CopyTagsToSnapshot": true,
          "DBInstanceClass": "db.t2.small",
          "DBInstanceIdentifier": "FarmerDB",
          "DBName": "FarmerDB",
          "DBSubnetGroupName": {
            "Ref": "FarmerDBSubnetGroup1B4DA139"
          },
          "DeletionProtection": false,
          "Engine": "mysql",
          "EngineVersion": "5.7.31",
          "MasterUserPassword": {
            "Fn::Join": [
              "",
              [
                "{{resolve:secretsmanager:",
                {
                  "Ref": "testFarmerDBSecretCBD37E0A3fdaad7efa858a3daf9490cf0a702aeb"
                },
                ":SecretString:password::}}"
              ]
            ]
          },
          "MasterUsername": "FarmerDBadmin",
          "MultiAZ": true,
          "Port": "3306",
          "PubliclyAccessible": false,
          "StorageEncrypted": true,
          "StorageType": "gp2",
          "VPCSecurityGroups": [
            {
              "Fn::GetAtt": [
                "FarmerDeliverySGDA50C27C",
                "GroupId"
              ]
            }
          ]
        },
        "Type": "AWS::RDS::DBInstance",
        "UpdateReplacePolicy": "Snapshot"
      },
      "FarmerDBSecretAttachment8D79E04A": {
        "Properties": {
          "SecretId": {
            "Ref": "testFarmerDBSecretCBD37E0A3fdaad7efa858a3daf9490cf0a702aeb"
          },
          "TargetId": {
            "Ref": "FarmerDBC89642C8"
          },
          "TargetType": "AWS::RDS::DBInstance"
        },
        "Type": "AWS::SecretsManager::SecretTargetAttachment"
      },
      "FarmerDBSubnetGroup1B4DA139": {
        "Properties": {
          "DBSubnetGroupDescription": "Subnet group for FarmerDB database",
          "SubnetIds": [
            {
              "Ref": "farmerVpcPublicSubnet1SubnetD688CBA4"
            },
            {
              "Ref": "farmerVpcPublicSubnet2Subnet9E651304"
            }
          ]
        },
        "Type": "AWS::RDS::DBSubnetGroup"
      },
      "FarmerDbProxy7A9720C2": {
        "Properties": {
          "Auth": [
            {
              "AuthScheme": "SECRETS",
              "IAMAuth": "DISABLED",
              "SecretArn": {
                "Ref": "FarmerDBSecretAttachment8D79E04A"
              }
            }
          ],
          "DBProxyName": "FarmerDbProxy",
          "EngineFamily": "MYSQL",
          "RequireTLS": true,
          "RoleArn": {
            "Fn::GetAtt": [
              "FarmerDbProxyIAMRoleF72268A5",
              "Arn"
            ]
          },
          "VpcSecurityGroupIds": [
            {
              "Fn::GetAtt": [
                "FarmerDeliverySGDA50C27C",
                "GroupId"
              ]
            }
          ],
          "VpcSubnetIds": [
            {
              "Ref": "farmerVpcPrivateSubnet1SubnetCCC9274B"
            },
            {
              "Ref": "farmerVpcPrivateSubnet2SubnetE7B0A714"
            }
          ]
        },
        "Type": "AWS::RDS::DBProxy"
      },
      "FarmerDbProxyIAMRoleDefaultPolicy212445D5": {
        "Properties": {
          "PolicyDocument": {
            "Statement": [
              {
                "Action": [
                  "secretsmanager:GetSecretValue",
                  "secretsmanager:DescribeSecret"
                ],
                "Effect": "Allow",
                "Resource": {
                  "Ref": "FarmerDBSecretAttachment8D79E04A"
                }
              }
            ],
            "Version": "2012-10-17"
          },
          "PolicyName": "FarmerDbProxyIAMRoleDefaultPolicy212445D5",
          "Roles": [
            {
              "Ref": "FarmerDbProxyIAMRoleF72268A5"
            }
          ]
        },
        "Type": "AWS::IAM::Policy"
      },
      "FarmerDbProxyIAMRoleF72268A5": {
        "Properties": {
          "AssumeRolePolicyDocument": {
            "Statement": [
              {
                "Action": "sts:AssumeRole",
                "Effect": "Allow",
                "Principal": {
                  "Service": "rds.amazonaws.com"
                }
              }
            ],
            "Version": "2012-10-17"
          }
        },
        "Type": "AWS::IAM::Role"
      },
      "FarmerDbProxyProxyTargetGroupEC553179": {
        "Properties": {
          "ConnectionPoolConfigurationInfo": {
            "ConnectionBorrowTimeout": 30,
            "MaxConnectionsPercent": 50
          },
          "DBInstanceIdentifiers": [
            {
              "Ref": "FarmerDBC89642C8"
            }
          ],
          "DBProxyName": {
            "Ref": "FarmerDbProxy7A9720C2"
          },
          "TargetGroupName": "default"
        },
        "Type": "AWS::RDS::DBProxyTargetGroup"
      },
      "FarmerDeliverySGDA50C27C": {
        "Properties": {
          "GroupDescription": "Shared SG for database and proxy",
          "SecurityGroupEgress": [
            {
              "CidrIp": "0.0.0.0/0",
              "Description": "Allow all outbound traffic by default",
              "IpProtocol": "-1"
            }
          ],
          "VpcId": {
            "Ref": "farmerVpc3E1EF418"
          }
        },
        "Type": "AWS::EC2::SecurityGroup"
      },
      "FarmerDeliverySGfromtestFarmerDeliverySG0E6AE860IndirectPort944272E1": {
        "Properties": {
          "Description": "Allow connections to the database Instance from the Proxy",
          "FromPort": {
            "Fn::GetAtt": [
              "FarmerDBC89642C8",
              "Endpoint.Port"
            ]
          },
          "GroupId": {
            "Fn::GetAtt": [
              "FarmerDeliverySGDA50C27C",
              "GroupId"
            ]
          },
          "IpProtocol": "tcp",
          "SourceSecurityGroupId": {
            "Fn::GetAtt": [
              "FarmerDeliverySGDA50C27C",
              "GroupId"
            ]
          },
          "ToPort": {
            "Fn::GetAtt": [
              "FarmerDBC89642C8",
              "Endpoint.Port"
            ]
          }
        },
        "Type": "AWS::EC2::SecurityGroupIngress"
      },
      "FarmerDeliverySGfromtestinboundDbAccessSecurityGroupFarmer80CA5D5033061FAB2892": {
        "Properties": {
          "Description": "from testinboundDbAccessSecurityGroupFarmer80CA5D50:3306",
          "FromPort": 3306,
          "GroupId": {
            "Fn::GetAtt": [
              "FarmerDeliverySGDA50C27C",
              "GroupId"
            ]
          },
          "IpProtocol": "tcp",
          "SourceSecurityGroupId": {
            "Fn::GetAtt": [
              "FarmerDeliverySGDA50C27C",
              "GroupId"
            ]
          },
          "ToPort": 3306
        },
        "Type": "AWS::EC2::SecurityGroupIngress"
      },
      "farmerVpc3E1EF418": {
        "Properties": {
          "CidrBlock": "10.0.0.0/16",
          "EnableDnsHostnames": true,
          "EnableDnsSupport": true,
          "InstanceTenancy": "default",
          "Tags": [
            {
              "Key": "Name",
              "Value": "test/farmer-Vpc"
            }
          ]
        },
        "Type": "AWS::EC2::VPC"
      },
      "farmerVpcIGW77EA3214": {
        "Properties": {
          "Tags": [
            {
              "Key": "Name",
              "Value": "test/farmer-Vpc"
            }
          ]
        },
        "Type": "AWS::EC2::InternetGateway"
      },
      "farmerVpcPrivateSubnet1DefaultRoute21094D32": {
        "Properties": {
          "DestinationCidrBlock": "0.0.0.0/0",
          "NatGatewayId": {
            "Ref": "farmerVpcPublicSubnet1NATGatewayAA1B3859"
          },
          "RouteTableId": {
            "Ref": "farmerVpcPrivateSubnet1RouteTableEA47B893"
          }
        },
        "Type": "AWS::EC2::Route"
      },
      "farmerVpcPrivateSubnet1RouteTableAssociationB94D34C5": {
        "Properties": {
          "RouteTableId": {
            "Ref": "farmerVpcPrivateSubnet1RouteTableEA47B893"
          },
          "SubnetId": {
            "Ref": "farmerVpcPrivateSubnet1SubnetCCC9274B"
          }
        },
        "Type": "AWS::EC2::SubnetRouteTableAssociation"
      },
      "farmerVpcPrivateSubnet1RouteTableEA47B893": {
        "Properties": {
          "Tags": [
            {
              "Key": "Name",
              "Value": "test/farmer-Vpc/PrivateSubnet1"
            }
          ],
          "VpcId": {
            "Ref": "farmerVpc3E1EF418"
          }
        },
        "Type": "AWS::EC2::RouteTable"
      },
      "farmerVpcPrivateSubnet1SubnetCCC9274B": {
        "Properties": {
          "AvailabilityZone": {
            "Fn::Select": [
              0,
              {
                "Fn::GetAZs": ""
              }
            ]
          },
          "CidrBlock": "10.0.128.0/18",
          "MapPublicIpOnLaunch": false,
          "Tags": [
            {
              "Key": "aws-cdk:subnet-name",
              "Value": "Private"
            },
            {
              "Key": "aws-cdk:subnet-type",
              "Value": "Private"
            },
            {
              "Key": "Name",
              "Value": "test/farmer-Vpc/PrivateSubnet1"
            }
          ],
          "VpcId": {
            "Ref": "farmerVpc3E1EF418"
          }
        },
        "Type": "AWS::EC2::Subnet"
      },
      "farmerVpcPrivateSubnet2DefaultRouteE813B3CB": {
        "Properties": {
          "DestinationCidrBlock": "0.0.0.0/0",
          "NatGatewayId": {
            "Ref": "farmerVpcPublicSubnet2NATGateway1A387200"
          },
          "RouteTableId": {
            "Ref": "farmerVpcPrivateSubnet2RouteTable859130FB"
          }
        },
        "Type": "AWS::EC2::Route"
      },
      "farmerVpcPrivateSubnet2RouteTable859130FB": {
        "Properties": {
          "Tags": [
            {
              "Key": "Name",
              "Value": "test/farmer-Vpc/PrivateSubnet2"
            }
          ],
          "VpcId": {
            "Ref": "farmerVpc3E1EF418"
          }
        },
        "Type": "AWS::EC2::RouteTable"
      },
      "farmerVpcPrivateSubnet2RouteTableAssociationB35F50E8": {
        "Properties": {
          "RouteTableId": {
            "Ref": "farmerVpcPrivateSubnet2RouteTable859130FB"
          },
          "SubnetId": {
            "Ref": "farmerVpcPrivateSubnet2SubnetE7B0A714"
          }
        },
        "Type": "AWS::EC2::SubnetRouteTableAssociation"
      },
      "farmerVpcPrivateSubnet2SubnetE7B0A714": {
        "Properties": {
          "AvailabilityZone": {
            "Fn::Select": [
              1,
              {
                "Fn::GetAZs": ""
              }
            ]
          },
          "CidrBlock": "10.0.192.0/18",
          "MapPublicIpOnLaunch": false,
          "Tags": [
            {
              "Key": "aws-cdk:subnet-name",
              "Value": "Private"
            },
            {
              "Key": "aws-cdk:subnet-type",
              "Value": "Private"
            },
            {
              "Key": "Name",
              "Value": "test/farmer-Vpc/PrivateSubnet2"
            }
          ],
          "VpcId": {
            "Ref": "farmerVpc3E1EF418"
          }
        },
        "Type": "AWS::EC2::Subnet"
      },
      "farmerVpcPublicSubnet1DefaultRouteD45A595C": {
        "DependsOn": [
          "farmerVpcVPCGW5B113563"
        ],
        "Properties": {
          "DestinationCidrBlock": "0.0.0.0/0",
          "GatewayId": {
            "Ref": "farmerVpcIGW77EA3214"
          },
          "RouteTableId": {
            "Ref": "farmerVpcPublicSubnet1RouteTable70D29438"
          }
        },
        "Type": "AWS::EC2::Route"
      },
      "farmerVpcPublicSubnet1EIPB227751A": {
        "Properties": {
          "Domain": "vpc",
          "Tags": [
            {
              "Key": "Name",
              "Value": "test/farmer-Vpc/PublicSubnet1"
            }
          ]
        },
        "Type": "AWS::EC2::EIP"
      },
      "farmerVpcPublicSubnet1NATGatewayAA1B3859": {
        "Properties": {
          "AllocationId": {
            "Fn::GetAtt": [
              "farmerVpcPublicSubnet1EIPB227751A",
              "AllocationId"
            ]
          },
          "SubnetId": {
            "Ref": "farmerVpcPublicSubnet1SubnetD688CBA4"
          },
          "Tags": [
            {
              "Key": "Name",
              "Value": "test/farmer-Vpc/PublicSubnet1"
            }
          ]
        },
        "Type": "AWS::EC2::NatGateway"
      },
      "farmerVpcPublicSubnet1RouteTable70D29438": {
        "Properties": {
          "Tags": [
            {
              "Key": "Name",
              "Value": "test/farmer-Vpc/PublicSubnet1"
            }
          ],
          "VpcId": {
            "Ref": "farmerVpc3E1EF418"
          }
        },
        "Type": "AWS::EC2::RouteTable"
      },
      "farmerVpcPublicSubnet1RouteTableAssociation8318E69F": {
        "Properties": {
          "RouteTableId": {
            "Ref": "farmerVpcPublicSubnet1RouteTable70D29438"
          },
          "SubnetId": {
            "Ref": "farmerVpcPublicSubnet1SubnetD688CBA4"
          }
        },
        "Type": "AWS::EC2::SubnetRouteTableAssociation"
      },
      "farmerVpcPublicSubnet1SubnetD688CBA4": {
        "Properties": {
          "AvailabilityZone": {
            "Fn::Select": [
              0,
              {
                "Fn::GetAZs": ""
              }
            ]
          },
          "CidrBlock": "10.0.0.0/18",
          "MapPublicIpOnLaunch": true,
          "Tags": [
            {
              "Key": "aws-cdk:subnet-name",
              "Value": "Public"
            },
            {
              "Key": "aws-cdk:subnet-type",
              "Value": "Public"
            },
            {
              "Key": "Name",
              "Value": "test/farmer-Vpc/PublicSubnet1"
            }
          ],
          "VpcId": {
            "Ref": "farmerVpc3E1EF418"
          }
        },
        "Type": "AWS::EC2::Subnet"
      },
      "farmerVpcPublicSubnet2DefaultRouteEE7A93B6": {
        "DependsOn": [
          "farmerVpcVPCGW5B113563"
        ],
        "Properties": {
          "DestinationCidrBlock": "0.0.0.0/0",
          "GatewayId": {
            "Ref": "farmerVpcIGW77EA3214"
          },
          "RouteTableId": {
            "Ref": "farmerVpcPublicSubnet2RouteTable3FCC446A"
          }
        },
        "Type": "AWS::EC2::Route"
      },
      "farmerVpcPublicSubnet2EIPFCBCD01A": {
        "Properties": {
          "Domain": "vpc",
          "Tags": [
            {
              "Key": "Name",
              "Value": "test/farmer-Vpc/PublicSubnet2"
            }
          ]
        },
        "Type": "AWS::EC2::EIP"
      },
      "farmerVpcPublicSubnet2NATGateway1A387200": {
        "Properties": {
          "AllocationId": {
            "Fn::GetAtt": [
              "farmerVpcPublicSubnet2EIPFCBCD01A",
              "AllocationId"
            ]
          },
          "SubnetId": {
            "Ref": "farmerVpcPublicSubnet2Subnet9E651304"
          },
          "Tags": [
            {
              "Key": "Name",
              "Value": "test/farmer-Vpc/PublicSubnet2"
            }
          ]
        },
        "Type": "AWS::EC2::NatGateway"
      },
      "farmerVpcPublicSubnet2RouteTable3FCC446A": {
        "Properties": {
          "Tags": [
            {
              "Key": "Name",
              "Value": "test/farmer-Vpc/PublicSubnet2"
            }
          ],
          "VpcId": {
            "Ref": "farmerVpc3E1EF418"
          }
        },
        "Type": "AWS::EC2::RouteTable"
      },
      "farmerVpcPublicSubnet2RouteTableAssociation0F0AFA60": {
        "Properties": {
          "RouteTableId": {
            "Ref": "farmerVpcPublicSubnet2RouteTable3FCC446A"
          },
          "SubnetId": {
            "Ref": "farmerVpcPublicSubnet2Subnet9E651304"
          }
        },
        "Type": "AWS::EC2::SubnetRouteTableAssociation"
      },
      "farmerVpcPublicSubnet2Subnet9E651304": {
        "Properties": {
          "AvailabilityZone": {
            "Fn::Select": [
              1,
              {
                "Fn::GetAZs": ""
              }
            ]
          },
          "CidrBlock": "10.0.64.0/18",
          "MapPublicIpOnLaunch": true,
          "Tags": [
            {
              "Key": "aws-cdk:subnet-name",
              "Value": "Public"
            },
            {
              "Key": "aws-cdk:subnet-type",
              "Value": "Public"
            },
            {
              "Key": "Name",
              "Value": "test/farmer-Vpc/PublicSubnet2"
            }
          ],
          "VpcId": {
            "Ref": "farmerVpc3E1EF418"
          }
        },
        "Type": "AWS::EC2::Subnet"
      },
      "farmerVpcVPCGW5B113563": {
        "Properties": {
          "InternetGatewayId": {
            "Ref": "farmerVpcIGW77EA3214"
          },
          "VpcId": {
            "Ref": "farmerVpc3E1EF418"
          }
        },
        "Type": "AWS::EC2::VPCGatewayAttachment"
      },
      "testFarmerDBSecretCBD37E0A3fdaad7efa858a3daf9490cf0a702aeb": {
        "Properties": {
          "Description": {
            "Fn::Join": [
              "",
              [
                "Generated by the CDK for stack: ",
                {
                  "Ref": "AWS::StackName"
                }
              ]
            ]
          },
          "GenerateSecretString": {
            "ExcludeCharacters": " %+~`#$&*()|[]{}:;<>?!'/@\"\\",
            "GenerateStringKey": "password",
            "PasswordLength": 30,
            "SecretStringTemplate": "{\"username\":\"FarmerDBadmin\"}"
          }
        },
        "Type": "AWS::SecretsManager::Secret"
      }
    }
  }
]